### DISCUSSION TITLE
Per-test System.Diagnostics.Activity, per-test logging, and the current poor Visual Studio Test Explorer and OTEL experience

### DISCUSSION THREAD

**Discussion by joshmouch** [xunit/xunit#2972] [Ideas]
Hello!

What are your thoughts on the xunit runners creating a new `System.Diagnostic.Activity` before each test method it executes, encapsulating the test in that `Activity`, then reverting back to the previous activity as the test completes?

There are two obvious reasons to do this:
First, Open Telemetry (OTEL) makes use of this `ActivitySource` to discover items to add to the telemetry.  So, say, I add an xunit project to a .NET Aspire Host (don't ask why, just an example), and in that nice dashboard, I've got the Structured tracing.  As of now, Xunit's tests would **not** show as items in that tracing.  

Simply doing something, like this, however, would then cause the tests to show up in OTEL:
```
[Fact]
public void TestThatWantsToBeSeen 
{
    var previousActivity = Activity.Current;  // This could be null. Or it could be set by some caller in the stack. (Maybe there's an xunit collection Activity, xunit class Activity, etc.) (Or maybe jetbrains adds one when using its test runner)
    Activity.Current = new Activity($"Running test {nameof(TestThatWantsToBeSeen)}");

    // Do my testing

    Activity.Current = previousActivity;
}

```

So now, let me switch topics really quick then come back.

I was just testing the MSTest vs Xunit v2 vs Xunit v3 Visual Studio Text Explorer experience, and MSTest was the only one to put the logs from the test into the output window.  I know there are workarounds, creating ILogger wrappers for ITestOutput and such, but MSTest just did it out of the box.  I didn't actually dig into how this was accomplished.

With that said, if I run the same Xunit v3 test from powershell "dotnet run" inside the test project folder, Xunit *does* show the exact same logging.  So, the logging is there.  I'm assuming this is using a different runner (console.runner instead of visualstudio.runner), and that hooks ITestOutput to the Console Automatically?

So then coming back:

The second reason is that Perhaps OTEL + Activity-per-test is a super easy way to get this nice Visual Studio Text Explorer user experience?  There'd still be some logging configuration, but that should be in a good state with a default project template.

Quick aside:
_I wrote my own plugin to do this.  It'd create an `Activity` for a `TestAssembly`, `TestCollection`, `TestClass`, `TestMethod`, etc.  It worked very well.  And an added bonus is that I was able to find performance problems *inside IClassFixtures* that normally don't report their time to anything I can easily see.  So I was able to shave minutes off each test class just by knowhere where, in a certain class fixture, there was a performance issue.
However, the xunit extensions I needed weren't all there or else I missed them, because it was waaaay more work than I initially thought it should be.  I ended up having to do xunit forks and whatnot.  And now I'm trying xunit.v3, so I'd have to start over... hence this discussion.

**Comment by bradwilson** [comment]
> With that said, if I run the same Xunit v3 test from powershell "dotnet run" inside the test project folder, Xunit does show the exact same logging. So, the logging is there. I'm assuming this is using a different runner (console.runner instead of visualstudio.runner), and that hooks ITestOutput to the Console Automatically?

That is correct.

Our VSTest adapter sends output back to VSTest: https://github.com/xunit/visualstudio.xunit/blob/9e3af28cec618b104e2f88159ca961d11c3fe450/src/xunit.runner.visualstudio/Sinks/VsExecutionSink.cs#L426-L427

We do not control what it does with it after that point. I have no idea what MSTest is doing that we're not doing, and more importantly, no idea whether what they're doing is something undocumented or not.

> The second reason is that Perhaps OTEL + Activity-per-test is a super easy way to get this nice Visual Studio Text Explorer user experience? There'd still be some logging configuration, but that should be in a good state with a default project template.

We don't control the default project template, either, so enabling this activity logging by default would involve convincing whatever Microsoft team that owns that to add whatever changes you'd need.

> What are your thoughts on the xunit runners creating a new System.Diagnostic.Activity before each test method it executes, encapsulating the test in that Activity, then reverting back to the previous activity as the test completes?

Given the [minimum target framework support for `System.Diagnostics.Activity`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity#applies-to), this would be something that we could theoretically add to v3, but definitely not to v2.

The [`TestContext`](https://github.com/xunit/xunit/blob/59c3e75edc229af1695a91d0bd1fd3dba51985c5/src/xunit.v3.core/TestContext.cs) seems like an ideal and simple place to add this for v3. We are manufacturing one of these for each phase of the test pipeline (assembly, collection, etc.), as well a making them for each stage of the phase (initializing, discovering, running, and cleaning up). That seems like a perfect match to what you're asking for, if I understand it correctly.

One thing for certain: I'd need to understand what the performance cost is. If we add activities, how much slower do things get when nobody is listening to them? How much slower when OTEL is listening and logging them? The NOOP scenario is extremely important for the feature in general, and the performance hit cost when there's a listener would be something the team that owns the default project template would want to know before they'd be willing to turn it on for everybody by default.

**Reply by bradwilson** [reply]
> Given the [minimum target framework support for System.Diagnostics.Activity](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity#applies-to), this would be something that we could theoretically add to v3, but definitely not to v2.

Worth mentioning that, a month later, we have released our first v3 prerelease packages, and:

> We don't control the default project template, either

At least for the moment, we do control the xunit3 templates (because they're not in-box).

> It'd create an Activity for a TestAssembly, TestCollection, TestClass, TestMethod, etc.

Thinking about this in the context of `TestContext`, it makes me wonder whether we could/should open up a hook for developers to be able to be notified when a new `TestContext` is being created. We don't currently dispose of them (they're not disposable, and we make them more frequently than just once per layer), but we do have a good idea about when they're going "out of scope" and that could theoretically be something we were able to surface if such a "`TestContext` lifetime" hook existed. It may not even be so much about the `TestContext` itself as it is about the start and end of each execution pipeline at each layer (so, being able to be notified that "test collection 'X' is starting", "test collection "X" has just finished", etc.
